File: laravel/laravel/app/Models/WasteReport.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\HasOne;

class WasteReport extends Model
{
    use HasFactory;

    // Status Constants
    public const STATUS_PENDING_VALIDATION = 'PENDING_VALIDATION';
    public const STATUS_VALIDATED = 'VALIDATED';
    public const STATUS_REJECTED = 'REJECTED';
    public const STATUS_ASSIGNED = 'ASSIGNED';
    public const STATUS_IN_PROGRESS = 'IN_PROGRESS';
    public const STATUS_PENDING_VERIFICATION = 'PENDING_VERIFICATION';
    public const STATUS_RESOLVED = 'RESOLVED';

    protected $fillable = [
        'report_code',
        'reported_by',
        'village_id',
        'category_id',
        'description',
        'status',
        'validated_by',
        'validated_at',
        'resolved_at',
        'location',
    ];

    protected function casts(): array
    {
        return [
            'validated_at' => 'datetime',
            'resolved_at' => 'datetime',
            'created_at' => 'datetime',
            'updated_at' => 'datetime',
        ];
    }

    public function reporter(): BelongsTo
    {
        return $this->belongsTo(User::class, 'reported_by');
    }

    public function village(): BelongsTo
    {
        return $this->belongsTo(Village::class, 'village_id');
    }

    public function category(): BelongsTo
    {
        return $this->belongsTo(WasteCategory::class, 'category_id');
    }

    public function validator(): BelongsTo
    {
        return $this->belongsTo(User::class, 'validated_by');
    }

    public function photos(): HasMany
    {
        return $this->hasMany(WasteReportPhoto::class, 'report_id');
    }

    public function statusHistories(): HasMany
    {
        return $this->hasMany(WasteReportStatusHistory::class, 'report_id');
    }

    public function cleanupTask(): HasOne
    {
        return $this->hasOne(CleanupTask::class, 'report_id');
    }

    protected $appends = ['status_label'];

    /**
     * Scope for active reports (reports that are still actively tracked on heatmaps).
     */
    public function scopeActive(Builder $query): Builder
    {
        return $query->whereNotIn('status', [self::STATUS_RESOLVED, self::STATUS_REJECTED]);
    }

    /**
     * Scope to automatically select latitude and longitude from PostGIS location point.
     */
    public function scopeWithCoordinates(Builder $query): Builder
    {
        if (empty($query->getQuery()->columns)) {
            $query->select('waste_reports.*');
        }

        return $query->selectRaw('ST_Y(waste_reports.location) as latitude, ST_X(waste_reports.location) as longitude');
    }

    public function getLatitudeAttribute($value)
    {
        if ($value !== null) {
            return (float) $value;
        }

        if (isset($this->attributes['location'])) {
            $point = \Illuminate\Support\Facades\DB::selectOne(
                'SELECT ST_Y(location) as lat FROM waste_reports WHERE id = ?',
                [$this->id]
            );
            return $point ? (float) $point->lat : null;
        }

        return null;
    }

    public function getLongitudeAttribute($value)
    {
        if ($value !== null) {
            return (float) $value;
        }

        if (isset($this->attributes['location'])) {
            $point = \Illuminate\Support\Facades\DB::selectOne(
                'SELECT ST_X(location) as lng FROM waste_reports WHERE id = ?',
                [$this->id]
            );
            return $point ? (float) $point->lng : null;
        }

        return null;
    }

    public function getStatusLabelAttribute(): ?string
    {
        if (empty($this->status)) {
            return null;
        }

        return match ($this->status) {
            self::STATUS_PENDING_VALIDATION => 'Menunggu Validasi',
            self::STATUS_VALIDATED => 'Tervalidasi',
            self::STATUS_REJECTED => 'Ditolak',
            self::STATUS_ASSIGNED => 'Ditugaskan',
            self::STATUS_IN_PROGRESS => 'Sedang Dibersihkan',
            self::STATUS_PENDING_VERIFICATION => 'Menunggu Verifikasi',
            self::STATUS_RESOLVED => 'Selesai',
            default => (string) $this->status,
        };
    }

    public function isPendingValidation(): bool
    {
        return $this->status === self::STATUS_PENDING_VALIDATION;
    }

    public function isValidated(): bool
    {
        return $this->status === self::STATUS_VALIDATED;
    }

    public function isRejected(): bool
    {
        return $this->status === self::STATUS_REJECTED;
    }

    public function isResolved(): bool
    {
        return $this->status === self::STATUS_RESOLVED;
    }
}

File: mobile-app/mobile/lib/models/waste_report.dart
import 'waste_category.dart';

class WasteReportPhoto {
  final int id;
  final String url;
  final bool isAfter;

  WasteReportPhoto({
    required this.id,
    required this.url,
    this.isAfter = false,
  });

  factory WasteReportPhoto.fromJson(Map<String, dynamic> json) {
    return WasteReportPhoto(
      id: json['id'] ?? 0,
      url: json['url'] ?? '',
      isAfter: json['type'] == 'AFTER' || json['is_after'] == true,
    );
  }
}

class StatusHistoryItem {
  final String fromStatus;
  final String toStatus;
  final String? notes;
  final String? changedByName;
  final DateTime? createdAt;

  StatusHistoryItem({
    required this.fromStatus,
    required this.toStatus,
    this.notes,
    this.changedByName,
    this.createdAt,
  });

  factory StatusHistoryItem.fromJson(Map<String, dynamic> json) {
    return StatusHistoryItem(
      fromStatus: json['from_status'] ?? '',
      toStatus: json['to_status'] ?? '',
      notes: json['notes'],
      changedByName: json['changed_by'] is Map ? json['changed_by']['name'] : null,
      createdAt: json['created_at'] != null ? DateTime.tryParse(json['created_at']) : null,
    );
  }
}

class WasteReport {
  final int id;
  final String reportCode;
  final String status;
  final String statusLabel;
  final String? description;
  final double? latitude;
  final double? longitude;
  final int? villageId;
  final String? villageName;
  final WasteCategory? category;
  final String? reporterName;
  final List<WasteReportPhoto> photos;
  final List<WasteReportPhoto> cleanupPhotos;
  final List<StatusHistoryItem> statusHistories;
  final DateTime? createdAt;
  final DateTime? resolvedAt;

  WasteReport({
    required this.id,
    required this.reportCode,
    required this.status,
    required this.statusLabel,
    this.description,
    this.latitude,
    this.longitude,
    this.villageId,
    this.villageName,
    this.category,
    this.reporterName,
    this.photos = const [],
    this.cleanupPhotos = const [],
    this.statusHistories = const [],
    this.createdAt,
    this.resolvedAt,
  });

  bool get isResolved => status == 'RESOLVED';
  bool get isPending => status == 'PENDING_VALIDATION';
  bool get isValidated => status == 'VALIDATED';
  bool get isInProgress => status == 'IN_PROGRESS' || status == 'ASSIGNED';
  bool get isRejected => status == 'REJECTED';

  String? get primaryPhotoUrl {
    if (photos.isNotEmpty) return photos.first.url;
    return null;
  }

  factory WasteReport.fromJson(Map<String, dynamic> json) {
    List<WasteReportPhoto> reportPhotos = [];
    if (json['photos'] is List) {
      reportPhotos = (json['photos'] as List)
          .map((p) => WasteReportPhoto.fromJson(p))
          .toList();
    } else if (json['primary_photo_url'] != null) {
      reportPhotos = [
        WasteReportPhoto(id: 0, url: json['primary_photo_url']),
      ];
    }

    List<WasteReportPhoto> taskPhotos = [];
    if (json['cleanup_task'] is Map && json['cleanup_task']['photos'] is List) {
      taskPhotos = (json['cleanup_task']['photos'] as List)
          .map((p) => WasteReportPhoto.fromJson(p))
          .toList();
    } else if (json['after_photos'] is List) {
      taskPhotos = (json['after_photos'] as List)
          .map((p) => WasteReportPhoto.fromJson(p))
          .toList();
    }

    List<StatusHistoryItem> histories = [];
    if (json['status_histories'] is List) {
      histories = (json['status_histories'] as List)
          .map((h) => StatusHistoryItem.fromJson(h))
          .toList();
    }

    WasteCategory? cat;
    if (json['category'] is Map) {
      cat = WasteCategory.fromJson(json['category']);
    } else if (json['category_name'] != null) {
      cat = WasteCategory(id: 0, name: json['category_name']);
    }

    String? vName = json['village_name'];
    int? vId = json['village_id'];
    if (json['village'] is Map) {
      vName = json['village']['name'] ?? vName;
      vId = json['village']['id'] ?? vId;
    }

    return WasteReport(
      id: json['id'] ?? 0,
      reportCode: json['report_code'] ?? '',
      status: json['status'] ?? 'PENDING_VALIDATION',
      statusLabel: json['status_label'] ?? json['status'] ?? '',
      description: json['description'],
      latitude: json['latitude'] != null ? (json['latitude'] as num).toDouble() : null,
      longitude: json['longitude'] != null ? (json['longitude'] as num).toDouble() : null,
      villageId: vId,
      villageName: vName,
      category: cat,
      reporterName: json['reporter'] is Map ? json['reporter']['name'] : json['reporter_name'],
      photos: reportPhotos,
      cleanupPhotos: taskPhotos,
      statusHistories: histories,
      createdAt: json['created_at'] != null ? DateTime.tryParse(json['created_at']) : null,
      resolvedAt: json['resolved_at'] != null ? DateTime.tryParse(json['resolved_at']) : null,
    );
  }
}
